azure-kusto-ingest/azure/kusto/ingest/V2/kusto_storage_uploader.py

```diff
@@ -0,0 +1,72 @@
+from uuid import UUID
+from typing import Optional, Dict, Union, IO, AnyStr
+
+from azure.storage.blob import BlobServiceClient
+from azure.kusto.data import KustoClient, KustoConnectionStringBuilder
+from azure.kusto.ingest._resource_manager import _ResourceManager
+from azure.kusto.ingest.descriptors import BlobDescriptor
+from azure.kusto.data.exceptions import KustoBlobError, KustoUploadError
+from azure.kusto.ingest.V2.blob_source import BlobSource
+from azure.kusto.ingest.V2.local_source import LocalSource
+
+
+class _KustoStorageUploader:
+    _MAX_RETRIES = 3
+    _SERVICE_CLIENT_TIMEOUT_SECONDS = 10 * 60
+
+    def __init__(self, kcsb: Union[str, KustoConnectionStringBuilder]):
+        if not isinstance(kcsb, KustoConnectionStringBuilder):
+            kcsb = KustoConnectionStringBuilder(kcsb)
+
+        self._connection_datasource = kcsb.data_source
+        self._resource_manager = _ResourceManager(KustoClient(kcsb))
+        self._proxy_dict: Optional[Dict[str, str]] = None
+
+    def upload_blob(
+        self,
+        blob_name: str,
+        stream: IO[AnyStr],
+        size: Optional[int] = None,
+        source_id: Union[str, UUID] = None,
+    ) -> "BlobDescriptor":
+        """
+        Uploads and transforms stream into a BlobDescriptor instance
+        :param str blob_name:
+        :param IO[AnyStr] stream: stream to be uploaded from
+        :param Optional[int] size:  estimated size of file if known
+        :param Union[str, uuid] source_id: source id
+        :return new BlobDescriptor instance
+        """
+
+        containers = self._resource_manager.get_containers()
+        retries_left = min(self._MAX_RETRIES, len(containers))
+        for container in containers:
+            succeeded = True
+            try:
+                blob_service = BlobServiceClient(container.account_uri, proxies=self._proxy_dict)
+                blob_client = blob_service.get_blob_client(container=container.object_name, blob=blob_name)
+                blob_client.upload_blob(data=stream, timeout=self._SERVICE_CLIENT_TIMEOUT_SECONDS)
+                return BlobDescriptor(blob_client.url, size, source_id)
+
+            except Exception as e:
+                succeeded = False
+                retries_left = retries_left - 1
+                if retries_left == 0:
+                    raise KustoBlobError(e)
+            finally:
+                self._resource_manager.report_resource_usage_result(container.storage_account_name, succeeded)
+
+    def close(self) -> None:
+        self._resource_manager.close()
+
+    def upload_local_source(self, local_source: LocalSource) -> BlobSource:
+        try:
+            local_stream = local_source.data()
```

Comment:
Its not alwyas a  stream
Actually this code shouldnt work currently if i understand correctly as data () can return different types